src/key_store/mod.rs

```diff
@@ -0,0 +1,192 @@
+//! A storage solution for cryptographic key material.
+//!
+//! This module provides access to the `KeyStore` struct, which manages the
+//! storage of `CredentialBundle` and `KeyPackageBundle` instances. The
+//! development of this module is tracked in #337, which also includes a
+//! roadmap.
+//!
+//! The current key store enables the storage of `CredentialBundle` instances,
+//! and grants access to `CredentialBundle` references via the
+//! `SignaturePublicKey` of the corresponding `Credential`. The
+//! `CredentialBundle` instances are only accessible via reference, so that if a
+//! `KeyStore` is used in multiple groups, those groups can also share a
+//! `CredentialBundle`.
```

Comment:
Generally, groups will only ever read from a `CredentialBundle`, so `RwLock` should allow parallel access to the same credential without issues. This leads to your question regarding updating/deleting of a credential. I don't think Credentials or CredentialBundles should ever be updated, but only deleted. On an MLS level, the credential is part of the KeyPackage and can thus only be "changed" (i.e. replaced by a fresh one) by doing an Update. As the API stands right now, this requires client activity to intiate it. 

So the lifecycle of a CredentialBundle would look as follows:
- it gets created
- the client decides to use it in one or more groups by creating KeyPackageBundles with that CredentialBundle and updating  the groups with these KeyPackageBundles.
- potentially, the client publishes some KeyPackageBundles created with the CredentialBundle and thus potentially joins new groups with them
- if the client wants to at some point delete the CredentialBundle, it first has to create a new one and replace KeyPackages that use them. This includes KeyPackages the client previously published, but also the ones that are in use in various groups. If the client wants to be totally safe, it has to wait until all KeyPackages that include the CredentialBundle have expired (as others might have downloaded them from the server before the update and still plan to use them).

This raises the question of who keeps track of how/where a CredentialBundle is used and when KeyPackageBundles that use it expire. If at some point the KeyStore is responsible for generation of _all_ of the KeyPackageBundles (not just the "init" ones as is now the case), it could keep track of expiration and use. But that's not the case just yet.

Reply:
I agree that there should be no modification but only delete and create.
> Generally, groups will only ever read from a CredentialBundle, so RwLock should allow parallel access to the same credential without issues.

Right, but my understanding of the locking you do here is that the entire hashmap gets locked. This way any write operation to the hashmap (in future this might also include any read from something that's permanently stored and then cached in the hashmap) would block a read. Or do I misunderstand the design?

> This raises the question of who keeps track of how/where a CredentialBundle is used and when KeyPackageBundles that use it expire. 

Right, I think this is the main question. I currently expect this to be the client or the key store. We don't have to decide right now but we should make sure the API here is good for any route we might want to take or can be easily upgraded. And we should document all this here.

Reply:
I've updated #337 to document the plans and also added a short "Future work" section to the end of the module documentation.